### PR TITLE
let the student take exam without proctoring if the  exam has no due …

### PR DESCRIPTION
**Pull request by ibrahimahmed443**
…date and course end date is passed

### CODE REVIEW

**Comment by ibrahimahmed443** [comment]
@douglashall @afzaledx. This uses: https://github.com/edx/edx-platform/pull/11767

File: edx_proctoring/api.py [hunk 1]
```diff
@@ -1770,6 +1770,20 @@ def get_student_view(user_id, course_id, content_id,
...
         return None
 
+    credit_service = get_runtime_service('credit')
+
+    # call service to get course end date.
+    credit_state = credit_service.get_credit_state(
+        user_id,
+        course_id,
+        return_course_info=True
+    )
```
Comment (douglashall) [inline edx_proctoring/api.py]: Nit: Put this on a single line.

File: edx_proctoring/api.py [hunk 2]
```diff
@@ -1798,11 +1812,15 @@ def get_student_view(user_id, course_id, content_id,
...
+        sub_view_func = _get_practice_exam_view
     elif is_proctored_exam:
-        return _get_proctored_exam_view(exam, context, exam_id, user_id, course_id)
+        sub_view_func = _get_proctored_exam_view
 
-    return None
+    if sub_view_func:
+        return sub_view_func(exam, context, exam_id, user_id, course_id)
+    else:
+        return None
```
Comment (douglashall) [inline edx_proctoring/api.py]: Nice cleanup!

File: edx_proctoring/api.py [hunk 3]
```diff
@@ -1770,6 +1770,20 @@ def get_student_view(user_id, course_id, content_id,
...
+    # call service to get course end date.
+    credit_state = credit_service.get_credit_state(
+        user_id,
+        course_id,
+        return_course_info=True
+    )
+
+    # if exam has no due date and course end date is passed, we'd still let the student take exam but without proctoring
+    end_date = credit_state.get('course_end_date', None)
+    if context.get('due_date') is None and has_due_date_passed(end_date):
```
Comment (douglashall) [inline edx_proctoring/api.py]: I think we need to add a test for this condition.

File: edx_proctoring/api.py [hunk 4]
```diff
@@ -1770,6 +1770,16 @@ def get_student_view(user_id, course_id, content_id,
...
 
+    credit_service = get_runtime_service('credit')
+
+    # call service to get course end date.
+    credit_state = credit_service.get_credit_state(user_id, course_id, return_course_info=True)
+
+    # if exam has no due date and course end date is passed, we'd still let the student take exam but without proctoring
+    end_date = credit_state.get('course_end_date', None)
+    if context.get('due_date') is None and has_due_date_passed(end_date):
+        return None
```
Comment (douglashall) [inline edx_proctoring/api.py]: Thinking about this more. I think we want to still allow the exam to be timed if the course end date has passed. We only want to skip proctoring if the course end data has passed. Can we move this check down below the exam lookup and only return None if the course end date has passed and it is a proctored or practice exam? Also, I don't think we need to check the exam due date here. If the course end date has passed that is enough to skip proctoring.

**Comment by douglashall** [comment]
It seems like we lost the previous fix for https://github.com/edx/edx-proctoring/pull/282/files#diff-c80ba566532963b4396830d546129f80R1782. I'm not seeing it removed in the diff and there are no conflicts. Can you rebase this branch?

**Comment by douglashall** [comment]
:+1:

**Pull request merged**